src/watch.ts

```diff
@@ -0,0 +1,237 @@
+ *   Catchup() will paginate through changes that we have missed.
+ *
+ *  --- Watch ?
+ *   Watch is "Long Polled. This mode is not paginated and the event system in CouchDB send
+ *     events as they arrive, which is super cool and reactive.
+ *   One gotcha those events arrive at the same rate wether you are watching the last seq or not.
+ *
+ *   Example:
+ *    listener A - up to date
+ *    listener B - few sequences behind.
+ *
+ *    Package C is updated.
+ *
+ *    Listener A receive update C
+ *    listener B receive update N.
+ *
+ *    Listener A is up to date again
+ *    listener B is still few sequences behind and will not receive any other event
+ *      until an other package is updated.
+ *      It will never be up to date because he receive event at the same pace
+ *      as they arrive in listener A, even if it's not the same package.
+ *
+ *
+ *  --- We could use catchup with a timeout between poll then?
+ *   Yes !
+ *   When we are catched up, we could await between poll and we will receive N changes.
+ *   But long-polling is more efficient in term of bandwidth and more reactive.
+ */
+async function run(
+  stateManager: StateManager,
+  mainIndex: SearchIndex
+): Promise<void> {
+  await stateManager.save({
+    stage: 'watch',
+  });
+
+  changesConsumer = createChangeConsumer(stateManager, mainIndex);
+
+  await catchup(stateManager);
+
+  log.info('🚀  Index is up to date, watch mode activated');
+
+  await watch(stateManager);
+
+  log.info('🚀  watch is done');
+}
+
+/**
+ * Loop through all changes that may have been missed.
+ */
+async function catchup(stateManager: StateManager): Promise<void> {
+  let hasCaughtUp: boolean = false;
+
+  while (!hasCaughtUp) {
+    loopStart = Date.now();
+
+    try {
+      const npmInfo = await npm.getInfo();
+      totalSequence = npmInfo.seq;
+
+      const { seq } = await stateManager.get();
+
+      log.info('🚀  Catchup: continue since sequence [%d]', seq);
+
+      // Get one chunk of changes from registry
+      const changes = await npm.getChanges({
+        since: seq,
+        limit: config.replicateConcurrency,
+        include_docs: true,
+      });
+
+      log.info(changes);
+
+      for (const change of changes.results) {
+        changesConsumer.push(change);
+      }
+      await changesConsumer.drain();
+
+      const newState = await stateManager.get();
+      if (newState.seq! >= totalSequence) {
+        hasCaughtUp = true;
+      }
+    } catch (err) {
+      sentry.report(err);
+    }
+  }
+}
+
+/**
+ * Active synchronous mode with Registry.
+ * Changes are polled with a keep-alived connection.
+ *
+ * @param stateManager - The state manager.
+ * @param mainIndex - Algolia index manager.
+ */
+async function watch(stateManager: StateManager): Promise<true> {
+  const { seq } = await stateManager.get();
+
+  const listener = npm.listenToChanges({
+    since: String(seq),
+    include_docs: false,
+    heartbeat: 30 * 1000,
+    // limit: 100,
```

Comment:
does this limit not come from config?

Reply:
actually it was wrong, there is no limit as we can only receive one per one